Create backend calls for returning the nextprevious paper

domienderle opened the pull request:
Closes #31 

## What I have made

I added three methods to:
 - go to the next paper (in review mode)
 - go to the next paper (not in review mode)
 - go to previous paper (not in review mode)

The next/previous papers are found according to issue #31.

## Checklist

- [x] I have updated the documentation accordingly and commented my code
- [x] (for reviewer) I have checked the implementation against the requirements

mowi12 reviewed (changes requested):
First, I think you need to update the description of the issue itself, including the acceptance criteria. Currently, you only state you want two API calls. That's not what the code is providing. And I think the wording is not correct, the API is already providing the interface, the mock-backend is simply not implementing it, or?

In addition, could you then specify which API call should have which functionality?
I already looked at the code, and what I saw looks good, but I want to compare the actual implementation against the wanted functionality from the issue.

domienderle reviewed (commented):
Yes, you are right. This was not the correct description of the issue. I actually wanted to remove the issue from ```to review``` again until I implemented the new functionality of the previous paper button in review mode. But now it is implemented. The queueing of the papers to get the previous paper in review mode is handled in the fronted. Due to that, we don't need the method ```getPreviousPaperToReview``` anymore. To remove this method would also bring a new issue in the API with it. To a matter of time I would propose to ignore this in the current PR and I will make a new issue in the backend and the API for it. What do you think about this?

mowi12 reviewed (changes requested):
Generally I approve that we leave the unused call and create a new issue to remove this all in both repos.
I would suggest to remove the implementation of the unused call in the meantime, there is no reason to keep the current implementation.

Next, when we handle the `getPreviousPaperToReview` is handled in the frontend, why don't you also handle the `nextToReview` also in the frontend?
I know that this would be more code, but are there good reasons to split these functions to local and back end?
For me, this does not feel right.

domienderle commented:
> Generally I approve that we leave the unused call and create a new issue to remove this all in both repos. I would suggest to remove the implementation of the unused call in the meantime, there is no reason to keep the current implementation.

You are right. There are two loc we have to leave to don't get a pipeline error because of unused code, I removed the rest of the implementation.

> Next, when we handle the `getPreviousPaperToReview` is handled in the frontend, why don't you also handle the `nextToReview` also in the frontend? I know that this would be more code, but are there good reasons to split these functions to local and back end? For me, this does not feel right.

I first implemented this in the frontend, but first, as you said, this is really ugly. Second, I wouldn't do so, because you have to use ```getAllProjectPapers``` in the frontend to implement this functionality, even though the user just wants to view one or two next papers. This would cause a lot of useless network traffic. Additionally, imo we should split the logic of the backend from the frontend. The ```getPreviousPaperToReview``` is just a push and pop of a list, but the ```getNextToReview``` is real logic, which we have implemented in the mock-backend (will implement in the backend) for every other use case. To wrap it up, I am no fan of this.

mowi12 commented on src/service.ts:
```diff
@@ -1362,4 +1361,116 @@ export const snowballRService: ISnowballR = {
+                return;
+            }
+            // If no paper with a higher local id on the max stage exists, that needs a review
+            if (stage == PROJECTS.get(projectId)!.maxStage && nextPapers.length == 0) {
+                callback({
+                    code: status.UNAVAILABLE,
+                    details: "No next project paper available.",
+                });
+                return;
+            }
```
Why do we return `UNAVAILABLE` here, but in the end `NOT_FOUND`?

I think you can completely remove this part of the code and always return `UNAVAILABLE` instead of `NOT_FOUND` after the while loop. Or not?

domienderle replied:
The reason why I did this is, that I need to distinguish the error in the frontend. If there is just no next Paper to review, but everything worked well, the ```nextPaper``` button has to be disabled. Otherwise, if there was an error due to anything else, I want the user to know this via a toast message. To implement this functionality, I have to deliver to different error messages. Imo this two are the best existing error status codes for this. If you have a better suggestion to achieve this I am open to it.

luca-schlecker replied:
Do not use `UNAVAILABE` please. See [the status code descriptions.](https://grpc.io/docs/guides/status-codes/#the-full-list-of-status-codes)

domienderle commented on src/service.ts:
```diff
@@ -1362,4 +1361,116 @@ export const snowballRService: ISnowballR = {
+                    parseInt(paper.localId) > parseInt(currentPaperId),
+            );
+            // If a paper to review exists on the same stage
+            if (nextPapers.length > 0) {
+                callback(null, addProjectPaperReviews(nextPapers[0]));
+                return;
+            }
```
Good call. I'm not totally sure, but this would make the call resistent against any order changes of the local ids at the moment or in future implementation. I think it is a good idea.

mowi12 commented on src/service.ts:
```diff
@@ -1393,4 +1392,115 @@ export const snowballRService: ISnowballR = {
+        call: ServerUnaryCall<Id, Project_Paper>,
+        callback: sendUnaryData<Project_Paper>,
+    ): void {
+        const { id } = call.request;
+        const projectId = id.split("-")[0];
+        const projectPapers = PROJECT_PROJECT_PAPERS.get(projectId)!
+            .map((ppp) => PROJECT_PAPERS.get(ppp)!)
+            .map(addProjectPaperReviews);
+        const projectPaper = projectPapers.find((pp) => pp.id === id);
+        if (!projectPaper) {
+            callback({
+                code: status.NOT_FOUND,
+                message: "paper-not-found",
+            });
+            return;
+        }
```
This part of the code is almost identical in all three calls. Maybe extract this part into a separate helper function into the utils.ts file.
This makes all three calls a lot shorter and better to read.

domienderle replied:
I would not retrieve the error call back, but with the rest you are right. I refactored the code ahead of the error callback.

mowi12 replied:
Ok

mowi12 commented on src/service.ts:
```diff
@@ -1393,4 +1392,115 @@ export const snowballRService: ISnowballR = {
+        }
+        callback({
+            code: status.NOT_FOUND,
+            message: "no-next-paper",
```
To be consistent, I think you should write a full sentence, similar to all the other calls. Applies to all three calls.

domienderle replied:
I guess you are right. The problem is, that I have to check on the full sentence in the frontend then, to distinguish between the errors, but since no other way worked (for example via metadata) for now it should be good enough.

mowi12 reviewed (changes requested):
I think we are almost done now. Only a small refactoring and an update to the message is needed imo.

mowi12 commented on src/util.ts:
```diff
@@ -213,3 +223,23 @@ export function makeResponseAuthMetadata(tokenPair: TokenPair): Metadata {
+
+/**
+ * Searches for the project paper with the given id, the according project and its project papers.
+ * @param paperId the id of the paper that should be found.
```
Please insert a new line between the description and the `@param` part : )

domienderle replied:
Good call.

mowi12 reviewed (changes requested):
Perfect! Only one last line and then we are done!
